File: mock-facility-management/src/main/java/de/vkb/dojo/es/facilityManagement/services/events/delegating/PickyEventAggregator.java
package de.vkb.dojo.es.facilityManagement.services.events.delegating;

import de.vkb.dojo.es.facilityManagement.model.event.Event;
import de.vkb.dojo.es.facilityManagement.services.events.EventAggregator;
import de.vkb.dojo.es.facilityManagement.services.events.EventAggregatorResult;

public abstract class PickyEventAggregator<PE extends Event, E extends PE, A> implements EventAggregator<PE, A> {
    private final Class<E> clazz;

    public PickyEventAggregator(Class<E> clazz) {
        this.clazz = clazz;
    }

    public Boolean canProcess(PE event) {
        return clazz.isAssignableFrom(event.getClass());
    }

    @Override
    public EventAggregatorResult<PE, A> process(PE event, A aggregate) {
        if (!canProcess(event)) {
            throw new IllegalArgumentException("cannot handle event of class " + event.getClass().getName());
        }
        //noinspection unchecked
        return (EventAggregatorResult<PE, A>) doProcess((E) event, aggregate);
    }

    protected abstract EventAggregatorResult<E, A> doProcess(E event, A aggregate);
}

File: mock-human-resources/src/main/kotlin/de/vkb/dojo/es/humanResources/services/events/delegating/PickyEventAggregator.kt
package de.vkb.dojo.es.humanResources.services.events.delegating

import de.vkb.dojo.es.humanResources.model.event.Event
import de.vkb.dojo.es.humanResources.services.events.EventAggregator
import de.vkb.dojo.es.humanResources.services.events.EventAggregatorResult

abstract class PickyEventAggregator<PE : Event, E : PE, A>(private val clazz: Class<E>) : EventAggregator<PE, A> {
    fun canProcess(event: PE): Boolean {
        return clazz.isAssignableFrom(event.javaClass)
    }

    override fun process(event: PE, aggregate: A?): EventAggregatorResult<PE, A> {
        require(canProcess(event)) { "cannot handle event of class " + event.javaClass.name }
        @Suppress("UNCHECKED_CAST")
        return doProcess(event as E, aggregate) as EventAggregatorResult<PE, A>
    }

    protected abstract fun doProcess(event: E, aggregate: A?): EventAggregatorResult<E, A>
}
